src/types.jl

```diff
@@ -387,21 +388,191 @@ function System(;
+                neighbor_finder=NoNeighborFinder(),
+                exchange_logger=ReplicaExchangeLogger(n_replicas),
+                replica_loggers=Tuple(() for _ in 1:n_replicas),
+                force_units=u"kJ * mol^-1 * nm^-1",
+                energy_units=u"kJ * mol^-1",
+                k=Unitful.k,
+                gpu_diff_safe=isa(coords, CuArray))
+    D = n_dimensions(boundary)
+    G = gpu_diff_safe
+    T = Molly.float_type(boundary)
+    CU = isa(coords, CuArray)
+    A = typeof(atoms)
+    AD = typeof(atoms_data)
+    PI = typeof(pairwise_inters)
+    SI = typeof(specific_inter_lists)
+    GI = typeof(general_inters)
+    C = typeof(coords)
+    B = typeof(boundary)
+    NF = typeof(neighbor_finder)
+    EL = typeof(exchange_logger)
+    F = typeof(force_units)
+    E = typeof(energy_units)
+    
+
+    if isnothing(replica_velocities)
+        if force_units == NoUnits
+            replica_velocities = [zero(coords) for _ in 1:n_replicas]
+        else
+            replica_velocities = [zero(coords) * u"ps^-1" for _ in 1:n_replicas]
+        end
+    end
+    V = typeof(replica_velocities[1])
+    if !all(y -> typeof(y) == V, replica_velocities)
+        throw(ArgumentError("The velocities for all the replicas are not of the same type."))
+    end
+
+    if !all(y -> y==replica_velocities[1], replica_velocities)
+        throw(ArgumentError("Some replicas have different number of velocities."))
+    end
+    if length(atoms) != length(coords)
+        throw(ArgumentError("There are $(length(atoms)) atoms but $(length(coords)) coordinates"))
+    end
+    if length(atoms) != length(replica_velocities[1])
+        throw(ArgumentError("There are $(length(atoms)) atoms but $(length(replica_velocities[1])) velocities"))
+    end
+    if length(atoms_data) > 0 && length(atoms) != length(atoms_data)
+        throw(ArgumentError("There are $(length(atoms)) atoms but $(length(atoms_data)) atom data entries"))
+    end
+
+    if isa(atoms, CuArray) && !isa(coords, CuArray)
+        throw(ArgumentError("The atoms are on the GPU but the coordinates are not"))
+    end
+    if isa(coords, CuArray) && !isa(atoms, CuArray)
+        throw(ArgumentError("The coordinates are on the GPU but the atoms are not"))
+    end
+
+    n_cuarray = sum(y -> isa(y, CuArray), replica_velocities)
+    if !(n_cuarray == n_replicas || n_cuarray == 0)
+        throw(ArgumentError("The velocities for $(n_cuarray) out of $(n_replicas) replicas are on GPU."))
+    end
+    if isa(atoms, CuArray) && n_cuarray != n_replicas
+        throw(ArgumentError("The atoms are on the GPU but the velocities are not"))
+    end
+    if n_cuarray == n_replicas && !isa(atoms, CuArray)
+        throw(ArgumentError("The velocities are on the GPU but the atoms are not"))
+    end
+
+    if length(replica_loggers) != n_replicas
+        throw(ArgumentError("There are $(length(replica_loggers)) loggers but $(n_replicas) replicas"))
+    end
+
+    if energy_units == NoUnits
```

Comment:
Since this code block appears in `System` too it can be moved to a function that takes `k` and `energy_units`.